### ISSUE TITLE
CS8604 false positive with indirect `string.IsNullOrEmpty()`

### ISSUE THREAD

**Issue by anilmujagic** [dotnet/roslyn#62147]
I use an extension method for `string.IsNullOrEmpty` for more concise and fluent code, which enables me to write `foobar.IsNullOrEmpty()` instead of `string.IsNullOrEmpty(foobar)`. However, compiler doesn't understand this and gives false positive warning/error.

This issue is basically the same as [CA1062 false-positives in FxCop](https://stackoverflow.com/questions/15188631/code-analysis-rule-ca1062-behaviour) I faced almost 10 years ago, but it's still present in the current C# compiler.

**Version Used**: .NET SDK 6.0.300, C# 10

**Steps to Reproduce**:

1. Create new console project:
    ```
    dotnet new console
    ```

2. Put following content in the `Program.cs` file:
    ```csharp
    public static class Program
    {
        public static bool IsNullOrEmpty(this string? str)
        {
            return string.IsNullOrEmpty(str);
        }

        public static void Main(string[] args)
        {
            var arg = args.Length > 0 ? "foo" : null;

            if (arg.IsNullOrEmpty() || "bar".Contains(arg))
                Console.WriteLine("baz");
        }
    }
    ```

3. Put following content in the `.csproj` file:
    ```xml
    <Project Sdk="Microsoft.NET.Sdk">

        <PropertyGroup>
            <OutputType>Exe</OutputType>
            <TargetFramework>net6.0</TargetFramework>
            <ImplicitUsings>enable</ImplicitUsings>
            <Nullable>enable</Nullable>
            <WarningsAsErrors>Nullable</WarningsAsErrors>
        </PropertyGroup>

    </Project>
    ```

4. Run it:
    ```
    dotnet run
    ```

Also available on [sharplab.io](https://sharplab.io/#v2:CYLg1APgxAdgrgGwQQwEYIKYAIMzZgWAChiABAJgEZizyBmYgbyxdZeLdLq1MoDYsqAPZCEWAJIBnAHKIEAeQBOAUQC2ABwAuATwAUmgBYBLST0oAGAPxZJmxQEoOrRk7YtSAdjPmAdFNlISmpaerYOANyuWAC+NESc3LwCpAAsWACyyEYwurzmANoAuljIigDmko7xzlFsAG6lJeVYALxNFT4AMrhlhlgAfFjmWNYARABmIqNYIFjwSJEk1W5YRuNYuqVlfjJyQRo6uvZYEBBYo6iloz4AwkIwmlkwkpvl9lUrK7wAnLoXyAAvUb2RZsWJEaJAA)

**Expected Behavior**:

Compile and run without errors.

**Actual Behavior**:

```
error CS8604: Possible null reference argument for parameter 'value' in 'bool string.Contains(string value)'. 
```

**Comment by Youssef1313** [comment]
You need to modify your code to be:

```csharp
#nullable enable

using System;
using System.Diagnostics.CodeAnalysis;

public static class Program
{
    // The NotNullWhen(false) tells the compiler "if this returned false, str is not null"
    public static bool IsNullOrEmpty([NotNullWhen(returnValue: false)] this string? str)
    {
        return string.IsNullOrEmpty(str);
    }

    public static void Main(string[] args)
    {
        var arg = args.Length > 0 ? "foo" : null;

        if (arg.IsNullOrEmpty() || "bar".Contains(arg))
            Console.WriteLine("baz");
    }
}
```

**Comment by anilmujagic** [comment]
@Youssef1313 Thank you!

**Comment by kondr1** [comment]
@anilmujagic Why you close it? @Youssef1313's solution working only for standard 2.1, net2.1 and above.
For others FTM it's still bug.
